rust/src/write.rs

```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
         Ok(Self {
             writer: Some(WriteMode::Raw(writer, chunk_mode)),
             options: opts,
-            schemas: HashMap::new(),
-            channels: HashMap::new(),
-            chunk_indexes: Vec::new(),
-            attachment_indexes: Vec::new(),
-            metadata_indexes: Vec::new(),
+            schemas: Default::default(),
+            channels: Default::default(),
+            chunk_indexes: Default::default(),
+            attachment_indexes: Default::default(),
+            metadata_indexes: Default::default(),
             message_bounds: None,
             channel_message_counts: BTreeMap::new(),
         })
     }
 
-    /// Adds a channel (and its provided schema, if any), returning its ID.
-    ///
-    /// Useful with subequent calls to [`write_to_known_channel()`](Self::write_to_known_channel)
-    pub fn add_channel(&mut self, chan: &Channel<'a>) -> McapResult<u16> {
-        let schema_id = match &chan.schema {
-            Some(s) => self.add_schema(s)?,
-            None => 0,
-        };
-
-        if let Some(id) = self.channels.get(chan) {
-            return Ok(*id);
+    // Adds a schema, returning its ID. If a schema with the same content has been added already,
+    // its ID is returned.
+    pub fn add_schema(&mut self, name: &str, encoding: &str, data: &[u8]) -> McapResult<u16> {
```

Comment:
Thoughts on having a separate struct that's like "ChannelWithoutId"? Pros: avoid having to thread multiple params through different layers, avoid possible confusion between params of the same type.